pyvista/examples/_example_loader.py

```diff
@@ -0,0 +1,307 @@
+"""Abstraction layer for downloading, reading, and loading example files."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+import functools
+import os
+from typing import (
+    Any,
+    Callable,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    runtime_checkable,
+)
+
+import pyvista
+from pyvista.core._typing_core import NumpyArray
+
+# The following classes define an API for working with either
+# a single filename or multiple filenames
+_FilenameType = TypeVar('_FilenameType', str, Tuple[str, ...], covariant=True)
+
+
+class _Filename(Protocol[_FilenameType]):
+    @property
+    @abstractmethod
+    def filename(self) -> _FilenameType: ...
+
+
+@runtime_checkable
+class _Downloadable(Protocol[_FilenameType]):
+    @abstractmethod
+    def download(self) -> _FilenameType: ...
+
+
+@runtime_checkable
+class _Loadable(Protocol):
+    @abstractmethod
+    def load(self) -> Any: ...
+
+
+class _MultiFilename(_Filename[Tuple[str, ...]]): ...
+
+
+class _SingleFilename(_Filename[str]): ...
+
+
+class _SingleFileLoadable(_SingleFilename, _Loadable):
+    """Wrap a single file for loading."""
+
+    def __init__(
+        self,
+        filename: str,
+        read_func: Optional[
+            Callable[[str], pyvista.DataSet | pyvista.Texture | NumpyArray[Any]]
+        ] = None,
+        load_func: Optional[Callable[[pyvista.DataSet], Any]] = None,
+    ):
+        # TODO: Refactor filename stuff into parent class
```

Comment:
Are you going to resolve TODO in this PR or keep it?

Reply:
Was just working on that, should be resolved by https://github.com/pyvista/pyvista/pull/5787/commits/9dcfa421f4c8dd4ce5445e42cfd7cde4b1b63dde